Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -4,34 +4,231 @@
 
 #define SUCCESS 0
 #define ERROR -1
-#define OVERFLOW -2
-#define EMPTY -3
+
+#define TEST 9
+#define EXPECTED 10
+
+#define SIZE 10
 
 // Hash struct
-typedef struct Hash {
+typedef struct Item {
     int key;
     int value;
-} Hash;
+} Item;
 
-void print() {
+//create global testHash and expectedHash for testFunctions to use
+Item* testHash[SIZE];
+Item* expectedHash[SIZE];
 
+// traverse function is for visual confirmation of differences between hashTables when building functionality. 
+// void traverse() {
+//     for (int i = 0; i < SIZE; i++) {
+//         int value = testHash[i]->value;
+//         int key = testHash[i]->key;        
+//         printf("Test Key: %d Value: %d \n", key, value);
+//     }
+//     for (int i = 0; i < SIZE; i++) {
+//         int value = expectedHash[i]->value;
+//         int key = expectedHash[i]->key;
+//         printf("Expected Key: %d Value: %d \n", key, value);
+//     }
+// }
+
+//compare checks the testHash and expectedHash for parity
+// if the two HashTables match returns true
+// if a mismatch occurs returns false
+bool compare() {
+    for (int i = 0; i < SIZE; i++) {
+        int testKey = testHash[i]->key;
+        int testValue = testHash[i]->value;
+        int expectedKey = expectedHash[i]->key;
+        int expectedValue = expectedHash[i]->value;
+        if (testKey != expectedKey || testValue != expectedValue) {
+            return false;
+        }
+    }
+    return true;
 }
 
-void traverse() {
+// hashing takes in a key
+// returns a "hashed" index based on the SIZE of the hashTable
+int hashing(int key) {
+    return key % SIZE;
+}
 
+// creates a new Item struct
+// returns new struct
+Item * createItem(int key, int value, int success) {
```

Comment:
Same here, `success` is passed in as a value.

Reply:
Resolved in [commit 12d1c07b69fa4b889c2b9b94f8721ec0c762256b](https://github.com/kcarrel/Training-Notes/pull/12/commits/12d1c07b69fa4b889c2b9b94f8721ec0c762256b).
Thought process: I was indeed meaning to pass a pointer to a success variable to Item * createItem as I didn't see another way to signify an ERROR or SUCCESS code like I normally would with an int function returning the appropriate variable. If there is another (or better) way of doing this I would love to be pointed in the direction of it! :)

Reply:
Another way to handle this could be to handle the error on the caller's side.
Something like:
```C
...
item = createItem(key,value)
if (item == NULL) {
  return ERROR
}
...
```

This is just stylistic, so just go with which ever approach you think is cleaner.

Reply:
Thanks Bernard!